Per-callback sampling

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21394 Per-callback sampling**

Summary:
Make it possible to use continuous logging (sampled callbacks) and
on-demand profiling (non-sampled callback) at the same time.

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

Differential Revision: [D15639723](https://our.internmc.facebook.com/intern/diff/D15639723)

Commit:
Per-callback sampling

Summary:
Make it possible to use continuous logging (sampled callbacks) and
on-demand profiling (non-sampled callback)

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

Commit:
Update on "Per-callback sampling"

Per-callback sampling

Summary:
Make it possible to use continuous logging (sampled callbacks) and
on-demand profiling (non-sampled callback)

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 21394 gh/ilia-cher/25/head

ilia-cher commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -8,20 +8,22 @@ namespace torch { namespace autograd { namespace profiler {
 std::vector<RecordFunctionCallback> start_callbacks;
 std::vector<RecordFunctionCallback> end_callbacks;
+std::vector<bool> is_callback_sampled;
+size_t num_sampled_callbacks = 0;
```
denormalized for perf reasons

ezyang commented:
I don't think I have enough context here to give a good review. Shout if you want me to look more closely.

suo commented:
Same :)

zdevito commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -8,20 +8,22 @@ namespace torch { namespace autograd { namespace profiler {
 }
 
 void setSamplingProbability(double prob) {
   if (std::abs(prob - 1.0) < kEps) {
```
If someone passes 1.0 it will be exactly equal. Why do we need this episilon stuff? It strikes me as cargo-culting not using floating point equality. But this is a comparison against 1.0, which is perfectly representable and users will just call setSamplingProbability(1.0) if they want it always on.

ilia-cher replied:
I was concerned about potential conversions (float/double), arithmetics and other ways an internal representation of 1.0 might potentially change to something not equal to 1.0 exactly, but if this is not an issue, I can simplify this to == with 1.0

zdevito reviewed (commented):
I can kinda see where this is going but the PR needs tests and context.

ilia-cher commented:
@zdevito please see the description of the PR for the context

Commit:
Update on "Per-callback sampling"

Per-callback sampling

Summary:
Make it possible to use continuous logging (sampled callbacks) and
on-demand profiling (non-sampled callback)

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 21394 gh/ilia-cher/25/head

ilia-cher commented:
@zdevito comment and test

dzhulgakov reviewed (commented):
Looks good, just one question

zdevito commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -30,19 +31,26 @@ double getSamplingProbability() {
+bool shouldRunSampledCallbacks() {
+  return (num_sampled_callbacks > 0) &&
+      (!sampling_prop_set ||
+      (((double) std::rand() / RAND_MAX) < sampling_prob));
```
It also has a really short period on windows (or at least it used to).

ilia-cher replied:
I could use thread_local mt19937

zdevito reviewed (approved):
This is fine with me. I'll let Dima make the final call on things here as I am pretty busy with other topics at the moment.

Commit:
Update on "Per-callback sampling"

Per-callback sampling

Summary:
Make it possible to use continuous logging (sampled callbacks) and
on-demand profiling (non-sampled callback)

Test Plan:
BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1 python setup.py develop
--cmake
./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

gh-metadata: pytorch pytorch 21394 gh/ilia-cher/25/head

ilia-cher commented:
using thread local PRNG
